### ISSUE TITLE
weights can't update

### ISSUE THREAD

**Issue by fhwzl** [pytorch/xla#2622]
env: pytorch 1.6 + torch_xla 1.6 + tensorflow114  modified based on self device backend

[mnist_training.py.txt](https://github.com/pytorch/xla/files/5522770/mnist_training.py.txt)
[test.log](https://github.com/pytorch/xla/files/5522773/test.log)

The problem is weights can't update.
Please help, and any  advise, thx

**Comment by zcain117** [comment]
Probably not the root cause, but 1.6 was our beta release and we now have a general availability release at 1.7

If you look at an [example training file](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py#L119), notice how the `xm.xla_device()` call happens inside the function that happens during multiprocessing, so that each process is working with 1 TPU core.

In your case, you are calling `xm.xla_device()` at the top-level of the python script. I'm not sure what would happen there, maybe it would just choose 1 device randomly? Or maybe it would result in the training not working like you're seeing.

Have you tried using the `torch_xla.distributed.xla_multiprocessing` module like shown in [this example](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist.py#L119)?

**Comment by jysohn23** [comment]
I see that the gradients are correct and equal between Aten CPU and XLA:TPU. Do you see the same behavior even if you increase the LR?

**Comment by zcain117** [comment]
Disregard my comment about `xm.xla_device()`. In theory, what you're doing should pick 1 random TPU core and should not cause backprop to fail

**Comment by fhwzl** [comment]
@jysohn23 set lr = 0.001, the weights also can't update.

**Comment by zcain117** [comment]
Could you temporarily try this for the __init__ model of the network:

```
  def __init__(self):
    super(MNIST, self).__init__()
    self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
    self.bn1 = nn.BatchNorm2d(10)
    self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
    self.bn2 = nn.BatchNorm2d(20)
    self.fc1 = nn.Linear(320, 50)
    self.fc2 = nn.Linear(50, 10)

  def forward(self, x):
    x = F.relu(F.max_pool2d(self.conv1(x), 2))
    x = self.bn1(x)
    x = F.relu(F.max_pool2d(self.conv2(x), 2))
    x = self.bn2(x)
    x = torch.flatten(x, 1)
    x = F.relu(self.fc1(x))
    x = self.fc2(x)
    return F.log_softmax(x, dim=1)
```

It's possible that one of the ops in your version have a bug where loss is not back propagating. We've seen this in the past once or twice ([example](https://github.com/pytorch/xla/issues/908)).

The version I pasted has daily tests on our side so it should not have any bug like that.

If this version of the network still doesn't work, then we know the issue is somewhere else

**Comment by fhwzl** [comment]
@zcain117 hi, I changed the test init mode under some limits, the test model like this:

```
    def __init__(self):
        super(MNIST, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.bn1 = nn.BatchNorm2d(10)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.bn2 = nn.BatchNorm2d(20)
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        """
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = self.bn1(x)
        x = F.relu(F.max_pool2d(self.conv2(x), 2))
        x = self.bn2(x)
        """
        x = F.relu(self.bn1(self.conv1(x)))
        x = F.max_pool2d(x, 2)
        x = F.relu(self.bn2(self.conv2(x)))
        x = F.max_pool2d(x, 2)
        x = torch.flatten(x, 1)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)
```

the py file and log file is:
[test.py.txt](https://github.com/pytorch/xla/files/5535100/test.py.txt)
[test.log](https://github.com/pytorch/xla/files/5535101/test.log)

weights also can't update

**Comment by zcain117** [comment]
I tried running this and don't see issues. Here it is with your exact code:
```
Epoch [1/5], Step [1/600], Loss cpu: 2.4581 / 2.4581, Loss dtu: 2.4578 / 2.4578
Epoch [1/5], Step [2/600], Loss cpu: 2.2625 / 2.2625, Loss dtu: 2.2617 / 2.2617
Epoch [1/5], Step [3/600], Loss cpu: 2.1052 / 2.1052, Loss dtu: 2.1041 / 2.1041
Epoch [1/5], Step [4/600], Loss cpu: 2.0466 / 2.0466, Loss dtu: 2.0458 / 2.0458
Epoch [1/5], Step [5/600], Loss cpu: 1.7795 / 1.7795, Loss dtu: 1.7787 / 1.7787
Epoch [1/5], Step [6/600], Loss cpu: 1.7739 / 1.7739, Loss dtu: 1.7736 / 1.7736
Epoch [1/5], Step [7/600], Loss cpu: 1.7007 / 1.7007, Loss dtu: 1.7000 / 1.7000
Epoch [1/5], Step [8/600], Loss cpu: 1.5438 / 1.5438, Loss dtu: 1.5430 / 1.5430
Epoch [1/5], Step [9/600], Loss cpu: 1.3825 / 1.3825, Loss dtu: 1.3814 / 1.3814
Epoch [1/5], Step [10/600], Loss cpu: 1.3796 / 1.3796, Loss dtu: 1.3790 / 1.3790
Epoch [1/5], Step [11/600], Loss cpu: 1.2932 / 1.2932, Loss dtu: 1.2927 / 1.2927
Epoch [1/5], Step [12/600], Loss cpu: 1.2532 / 1.2532, Loss dtu: 1.2525 / 1.2525
Epoch [1/5], Step [13/600], Loss cpu: 1.3080 / 1.3080, Loss dtu: 1.3077 / 1.3077
Epoch [1/5], Step [14/600], Loss cpu: 1.0942 / 1.0942, Loss dtu: 1.0935 / 1.0935
Epoch [1/5], Step [15/600], Loss cpu: 0.9991 / 0.9991, Loss dtu: 0.9974 / 0.9974
Epoch [1/5], Step [16/600], Loss cpu: 1.0362 / 1.0362, Loss dtu: 1.0359 / 1.0359
Epoch [1/5], Step [17/600], Loss cpu: 0.8373 / 0.8373, Loss dtu: 0.8370 / 0.8370
Epoch [1/5], Step [18/600], Loss cpu: 0.7678 / 0.7678, Loss dtu: 0.7682 / 0.7682
Epoch [1/5], Step [19/600], Loss cpu: 0.7718 / 0.7718, Loss dtu: 0.7697 / 0.7697
Epoch [1/5], Step [20/600], Loss cpu: 0.8505 / 0.8505, Loss dtu: 0.8498 / 0.8498
Epoch [1/5], Step [21/600], Loss cpu: 0.8683 / 0.8683, Loss dtu: 0.8667 / 0.8667
Epoch [1/5], Step [22/600], Loss cpu: 0.6083 / 0.6083, Loss dtu: 0.6073 / 0.6073
Epoch [1/5], Step [23/600], Loss cpu: 0.6480 / 0.6480, Loss dtu: 0.6473 / 0.6473
Epoch [1/5], Step [24/600], Loss cpu: 0.6366 / 0.6366, Loss dtu: 0.6373 / 0.6373
Epoch [1/5], Step [25/600], Loss cpu: 0.7027 / 0.7027, Loss dtu: 0.7024 / 0.7024
Epoch [1/5], Step [26/600], Loss cpu: 0.6308 / 0.6308, Loss dtu: 0.6310 / 0.6310
Epoch [1/5], Step [27/600], Loss cpu: 0.6323 / 0.6323, Loss dtu: 0.6312 / 0.6312
Epoch [1/5], Step [28/600], Loss cpu: 0.6977 / 0.6977, Loss dtu: 0.6965 / 0.6965
Epoch [1/5], Step [29/600], Loss cpu: 0.5164 / 0.5164, Loss dtu: 0.5164 / 0.5164
Epoch [1/5], Step [30/600], Loss cpu: 0.5660 / 0.5660, Loss dtu: 0.5653 / 0.5653
Epoch [1/5], Step [31/600], Loss cpu: 0.6620 / 0.6620, Loss dtu: 0.6609 / 0.6609
Epoch [1/5], Step [32/600], Loss cpu: 0.4708 / 0.4708, Loss dtu: 0.4707 / 0.4707
Epoch [1/5], Step [33/600], Loss cpu: 0.4995 / 0.4995, Loss dtu: 0.4981 / 0.4981
Epoch [1/5], Step [34/600], Loss cpu: 0.5287 / 0.5287, Loss dtu: 0.5285 / 0.5285
Epoch [1/5], Step [35/600], Loss cpu: 0.4769 / 0.4769, Loss dtu: 0.4768 / 0.4768
...
```

Loss goes down for both cpu and tpu but then it crashed when doing eval due to the `retain_grad`

I disabled the `retain_grad` and I see this:
```
Epoch [1/5], Step [100/600], Loss cpu: 0.1150 / 0.1150, Loss dtu: 0.1153 / 0.1153
Epoch [1/5], Step [200/600], Loss cpu: 0.0931 / 0.0931, Loss dtu: 0.0930 / 0.0930
Epoch [1/5], Step [300/600], Loss cpu: 0.1324 / 0.1324, Loss dtu: 0.1328 / 0.1328
Epoch [1/5], Step [400/600], Loss cpu: 0.0803 / 0.0803, Loss dtu: 0.0789 / 0.0789
Epoch [1/5], Step [500/600], Loss cpu: 0.0995 / 0.0995, Loss dtu: 0.1004 / 0.1004
Epoch [1/5], Step [600/600], Loss cpu: 0.1879 / 0.1879, Loss dtu: 0.1883 / 0.1883
Epoch [2/5], Step [100/600], Loss cpu: 0.0294 / 0.0294, Loss dtu: 0.0302 / 0.0302
Epoch [2/5], Step [200/600], Loss cpu: 0.0353 / 0.0353, Loss dtu: 0.0356 / 0.0356
Epoch [2/5], Step [300/600], Loss cpu: 0.0663 / 0.0663, Loss dtu: 0.0667 / 0.0667
Epoch [2/5], Step [400/600], Loss cpu: 0.0384 / 0.0384, Loss dtu: 0.0372 / 0.0372
Epoch [2/5], Step [500/600], Loss cpu: 0.0564 / 0.0564, Loss dtu: 0.0540 / 0.0540
Epoch [2/5], Step [600/600], Loss cpu: 0.1918 / 0.1918, Loss dtu: 0.1921 / 0.1921
Epoch [3/5], Step [100/600], Loss cpu: 0.0163 / 0.0163, Loss dtu: 0.0188 / 0.0188
Epoch [3/5], Step [200/600], Loss cpu: 0.0255 / 0.0255, Loss dtu: 0.0244 / 0.0244
Epoch [3/5], Step [300/600], Loss cpu: 0.0436 / 0.0436, Loss dtu: 0.0434 / 0.0434
Epoch [3/5], Step [400/600], Loss cpu: 0.0273 / 0.0273, Loss dtu: 0.0289 / 0.0289
Epoch [3/5], Step [500/600], Loss cpu: 0.0401 / 0.0401, Loss dtu: 0.0411 / 0.0411
Epoch [3/5], Step [600/600], Loss cpu: 0.1968 / 0.1968, Loss dtu: 0.1951 / 0.1951
Epoch [4/5], Step [100/600], Loss cpu: 0.0100 / 0.0100, Loss dtu: 0.0105 / 0.0105
Epoch [4/5], Step [200/600], Loss cpu: 0.0162 / 0.0162, Loss dtu: 0.0154 / 0.0154
Epoch [4/5], Step [300/600], Loss cpu: 0.0290 / 0.0290, Loss dtu: 0.0284 / 0.0284
Epoch [4/5], Step [400/600], Loss cpu: 0.0207 / 0.0207, Loss dtu: 0.0210 / 0.0210
Epoch [4/5], Step [500/600], Loss cpu: 0.0377 / 0.0377, Loss dtu: 0.0401 / 0.0401
Epoch [4/5], Step [600/600], Loss cpu: 0.1974 / 0.1974, Loss dtu: 0.1991 / 0.1991
Epoch [5/5], Step [100/600], Loss cpu: 0.0066 / 0.0066, Loss dtu: 0.0069 / 0.0069
Epoch [5/5], Step [200/600], Loss cpu: 0.0113 / 0.0113, Loss dtu: 0.0103 / 0.0103
Epoch [5/5], Step [300/600], Loss cpu: 0.0271 / 0.0271, Loss dtu: 0.0214 / 0.0214
Epoch [5/5], Step [400/600], Loss cpu: 0.0134 / 0.0134, Loss dtu: 0.0110 / 0.0110
Epoch [5/5], Step [500/600], Loss cpu: 0.0406 / 0.0406, Loss dtu: 0.0419 / 0.0419
Epoch [5/5], Step [600/600], Loss cpu: 0.1957 / 0.1957, Loss dtu: 0.1962 / 0.1962
Test Accuracy of the model on the 10000 test images: cpu 98.91 %, dtu 98.92 %
(torch-xla-nightly) zcain@zcain-vm:~$ 
```

Accuracy looks pretty good for 5 epochs

Do note that your script is only using 1 TPU core but the TPU comes with 8 cores. We have examples of the multiprocessing API like [this one](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py)

**Comment by fhwzl** [comment]
I think it's my problem of modifying the source code for lower version tensorflow. When use torch 1.2 + torch_xla 0.1 + tensorflow 114, It's OK. The simple example:
[test_xla.py.txt](https://github.com/pytorch/xla/files/5544564/test_xla.py.txt)
[test_xla_0_1.log](https://github.com/pytorch/xla/files/5544565/test_xla_0_1.log)
[test_xla_1_6.log](https://github.com/pytorch/xla/files/5544566/test_xla_1_6.log)
    torch_xla 0.1 (code changed for self device tensorflow backend) is ok with weights update.
    torch_xla 1.6 (code changed too) failed with updating weights.

**Comment by fhwzl** [comment]
The same test code, I changed `xm.optimizer_step(dtu_optimizer, barrier=True)` to `xm.optimizer_step(dtu_optimizer, barrier=False)`. 

[test_xla_1_6_.log](https://github.com/pytorch/xla/files/5546078/test_xla_1_6_.log)

weights seem to update, but with other problem in large models.  what is the most likely cause that weights can't update, thx

**Comment by zcain117** [comment]
I don't understand. The model you gave runs fine on CPU and TPU, loss is the same on both CPU and TPU, accuracy is the same on both CPU and TPU. If weights weren't updating, loss would not go down and accuracy would be random. So it seems like the weights are updating

**Comment by fhwzl** [comment]
I try torch1.7 + torch-xla0.1 + tf114 ok, weights can update. thx and close

**Comment by zcain117** [comment]
More details on the `barrier` arg: https://pytorch.org/xla/release/1.7/index.html#torch_xla.core.xla_model.optimizer_step

`barrier (bool, optional) – Whether the XLA tensor barrier should be issued in this API. If using the PyTorch XLA ParallelLoader or DataParallel support, this is not necessary as the barrier will be issued by the XLA data loader iterator next() call. Default: False`

**Comment by fhwzl** [comment]
I got it, thx again.
